pythia/models/butd.py

```diff
@@ -0,0 +1,294 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+
+import gc
+import torch
+from pythia.common.registry import registry
+from pythia.modules.encoders import ImageEncoder
+from pythia.modules.decoders import CaptioningDecoder
+from torch import nn
+
+from .base_model import BaseModel
+
+
+@registry.register_model("butd")
+class BUTD(BaseModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.config = config
+        self._global_config = registry.get("config")
+        self._datasets = self._global_config.datasets.split(",")
+
+    def build(self):
+        self._build_word_embedding()
+        self._init_feature_encoders("image")
+        self._init_decoder()
+        self._init_extras()
+
+    def _build_word_embedding(self):
+        self.text_processor = registry.get(self._datasets[0] + "_text_processor")
+        self.vocab = self.text_processor.vocab
+        self.vocab_size = self.vocab.get_size()
+        self.word_embedding = self.vocab.get_embedding(
+            torch.nn.Embedding, embedding_dim=300
+        )
+
+    def _init_feature_encoders(self, attr):
+        feat_encoders = []
+        feat_encoders_list_config = self.config[
+            attr + "_feature_encodings"
+        ]  # finetune_faster_rcnn_fpn_fc7
+        feature_dim = self.config[attr + "_feature_dim"]  # 2048
+        setattr(self, attr + "_feature_dim", feature_dim)
+
+        for feat_encoder in feat_encoders_list_config:
+            encoder_type = feat_encoder["type"]
+            encoder_kwargs = feat_encoder["params"]
+            encoder_kwargs["model_data_dir"] = self.config["model_data_dir"]
+
+            feat_model = ImageEncoder(encoder_type, feature_dim, **encoder_kwargs)
+
+            feat_encoders.append(feat_model)
+            setattr(self, attr + "_feature_dim", feat_model.out_dim)
+
+        setattr(self, attr + "_feature_encoders", nn.ModuleList(feat_encoders))
+
+    def _init_decoder(self):
+        self.decoder = CaptioningDecoder(
+            attention_dim=1024,
+            embed_dim=300,
+            decoder_dim=1024,
+            vocab_size=self.vocab_size,
+        )
+
+    def _init_extras(self):
+        self.inter_model = None
+
+    def get_optimizer_parameters(self, config):
+        params = [
+            {"params": self.word_embedding.parameters()},
+            {"params": self.decoder.parameters()},
+            {
+                "params": self.image_feature_encoders.parameters(),
+                "lr": (config["optimizer_attributes"]["params"]["lr"] * 0.1),
```

Comment:
I might be missing a detail here but why is the `lr` being multiplied by 0.1 here?

Reply:
This layer comes from faster-rcnn and is being finetuned. That's why LR for fine-tuning is 0.1 times lower than the normal LR.

Reply:
Maybe move this value also to `config`, 0.1 here looks like a magic number.